driver-core/src/main/java/com/datastax/driver/core/ControlConnection.java

```diff
@@ -568,7 +568,11 @@ private static void refreshNodeListAndTokenMap(Connection connection, Cluster.Ma
                 // We don't know that node, create the Host object but wait until we've set the known
                 // info before signaling the addition.
                 host = cluster.metadata.add(foundHosts.get(i));
-                isNew = true;
+                if (host == null)
+                    // There must have been a race condition and someone else added the node before us
+                    host = cluster.metadata.getHost(foundHosts.get(i));
```

Comment:
Yes, your right. Doing it this way would close the last small hole and it should be easy enough to do. I have an updated patch up soon.